[hop] local_map validate partitioned fw/bw wrt placements

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #164437
* #164433
* #164431
* #163602
* #164340
* __->__ #164420
* #164419
* #164321
* #164296

Reviewed GPT-5 Summary:

**Summary / Goal**  
Add validation that partitioned forward/backward graphs respect placements.  

**Details**  
- Validates placement alignment in local_map.  
- The HOP's autograd key gets called when we are tracing the joint, we need to validate:
  - the inputs to the HOP's fwd gm (typically this is the dynamo rewritten inputs)
  - the inputs to the HOP partitioned fwd/bwd gm
  - the outputs of the HOP partitioned fwd/bwd gm

**Motivation**  
Catch mismatch errors earlier, improve debugging.

xmfan committed:
[hop] local_map validate partitioned fw/bw wrt placements

[ghstack-poisoned]

ezyang commented on torch/_higher_order_ops/local_map.py:
```diff
@@ -194,6 +198,40 @@ def fw_with_masks(*args: Any) -> tuple[tuple[Any], list[bool]]:
+        assert len(new_bw_gm.graph.find_nodes(op="output")) == 1
+        actual_bw_outputs = len(new_bw_gm.graph.find_nodes(op="output")[0].args[0])
+        assert expected_bw_inputs == actual_bw_inputs
+        assert expected_bw_outputs == actual_bw_outputs
```
for these asserts, think about what kind of extra info you wanna see and have the assert messages print them. Could ask llm to do this

xmfan committed:
Update on "[hop] local_map validate partitioned fw/bw wrt placements"


Reviewed GPT-5 Summary:

**Summary / Goal**  
Add validation that partitioned forward/backward graphs respect placements.  

**Details**  
- Validates placement alignment in local_map.  
- The HOP's autograd key gets called when we are tracing the joint, we need to validate:
  - the inputs to the HOP's fwd gm (typically this is the dynamo rewritten inputs)
  - the inputs to the HOP partitioned fwd/bwd gm
  - the outputs of the HOP partitioned fwd/bwd gm

**Motivation**  
Catch mismatch errors earlier, improve debugging.  

[ghstack-poisoned]
